docs_src/microservices/security/Ch-SSH-Tunneling-HowToSecureDeviceServices.md

```diff
@@ -0,0 +1,297 @@
+
+```sh
+ssh-keygen -q -t rsa -C root -N '' -f ~/.ssh/id_rsa 2>/dev/null
+```
+
+One example of build script to make SSH public key be injected is given as like:
+
+[example build shell script to build and install SSH public key into sshd server](https://github.com/edgexfoundry-holding/external-device-security-examples/ssh-tunneling/build.sh)
+
+### SSH Tunneling: Local Port Forwarding
+
+from primary to secondary /remote
+it is achieved by -L flags of ssh command.
+
+```sh
+ssh -vv -o StrictHostKeyChecking=no \
+  -o UserKnownHostsFile=/dev/null \
+  -N $TUNNEL_HOST \
+  -L *:$SERVICE_PORT:$SERVICE_HOST:$SERVICE_PORT \
+  -p $TUNNEL_SSH_PROT
+```
+
+where environment variables are:
+
+- TUNNEL_HOST is the remote host name or IP address that SSH daemon or server is running on;
+
+- TUNNEL_SSH_PROT is the port number to be used on the SSH tunnel communication between the primary machine and the remote machine
+
+- SERVICE_PORT is the port number from the local or the primary to be forwared to the remote machine; without lose of generality, the port number on the remote machine is the same as the primary one
+
+- SERVICE_HOST is the service host name or IP address of the Docker containers that are running on the remote machine;
+
+### SSH Reverse Tunneling: Remote Port Forwarding
+
+the reverse direction: from the secondary /remote back to the primary
+
+The reverse SSH tunneling is also needed because the device services depends on the core services like `data`, `metadata`, ... and so on.  These core services are running on the primray machine and should be **reversely** tunneling back to the device services on the remote side through the SSH remote port forwarding connection.  This can be achieved by using `-R` flag of ssh command.
+
+```sh
+ssh -vv -o StrictHostKeyChecking=no \
+  -o UserKnownHostsFile=/dev/null \
+  -N $TUNNEL_HOST \
+  -R 0.0.0.0:48080:edgex-core-data:48080 \
+  -R 0.0.0.0:5563:edgex-core-data:5563 \
+  -R 0.0.0.0:48081:edgex-core-metadata:48081 \
+  -R 0.0.0.0:8500:edgex-core-consul:8500 \
+  -p $TUNNEL_SSH_PROT
+```
+
+where environment variables are:
+
+- TUNNEL_HOST is the remote host name or IP address that SSH daemon or server is running on;
+
+In the reverse tunneling, the service host names of dependent services are used like `edgex-core-data`, for example.
+
+### Put it all together
+
+- Launch the remote machine or VM if it is not yet:
+
+```sh
+~/vm/vagrant up
+```
+
+and ssh into the remote machine via `~/vm/vagrant ssh`
+
+- In the primary machine, generate ssh key pairs using ssh-keygen:
+
+```sh
+ssh-keygen -q -t rsa -C root -N '' -f ~/.ssh/id_rsa 2>/dev/null
+```
+
+This produces two files under directory ~/.ssh: one for private key (id_rsa) and one for public key (id_rsa.pub)
+
+- Make docker build with ssh-device-proxy Docker file and entrypoint.sh:
```

Comment:
Is this supposed to be a command? If so should say "Run `make docker build` ..."

Docker file is one word "Dockerfile"
back tick around `entrypoint.sh`